File: app/servicos/documento.py
"""Geracao do parecer tecnico em .docx a partir do modelo docxtpl.

O parecer e um .docx de layout fixo com tabelas de celulas mescladas, cabecalho
MEC/UFVJM com brasao e rodape de duas colunas. So o conteudo varia - o
cabecalho NUNCA e recriado por codigo.
"""

from __future__ import annotations

import hashlib
import unicodedata
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path

from docx import Document
from docx.shared import Mm
from docxtpl import DocxTemplate, InlineImage, RichText

from app.config import RAIZ, obter_config
from app.servicos import datas_br, textos

DIR_MODELOS = RAIZ / "app" / "templates"
MODELO_V1 = "modelo_parecer_v1.docx"
MODELO_V2 = "modelo_parecer_v2.docx"

CHAVES_RICHTEXT = (
    "posto_trabalho",
    "agente_nocivo",
    "fundamentacao_legal",
    "alteracao",
    "recomendacao_tecnica",
    "reavaliacao",
)


class DadosIncompletos(ValueError):
    """RN-04 - a emissao lista o que falta, nunca falha em silencio."""

    def __init__(self, faltantes: list[str]):
        self.faltantes = faltantes
        super().__init__("Faltam dados obrigatorios: " + "; ".join(faltantes))


class ExposicoesDivergentes(ValueError):
    pass


def sha256_arquivo(caminho: Path) -> str:
    h = hashlib.sha256()
    with caminho.open("rb") as f:
        for bloco in iter(lambda: f.read(65536), b""):
            h.update(bloco)
    return h.hexdigest()


def sha256_texto(texto: str) -> str:
    return hashlib.sha256(texto.encode("utf-8")).hexdigest()


def caminho_modelo(nome: str | None) -> Path:
    return DIR_MODELOS / (nome or MODELO_V1)


def _rich(valor: str | None) -> RichText | None:
    """Quebras de linha em celula: RichText com \\a e {{r chave }}."""
    if valor is None:
        return None
    texto = str(valor).replace("\r\n", "\n").rstrip()
    if texto == "":
        return None
    rt = RichText()
    for i, linha in enumerate(texto.split("\n")):
        if i:
            rt.add("\a")
        rt.add(linha)
    return rt


def rich_multilinha(valor: str | None) -> RichText | None:
    """`_rich` para os outros documentos do sistema.

    Existe porque a lista de presenca (fatia 3 de Certificados) precisa da mesma
    quebra de linha dentro de celula, e importar um `_privado` de outro modulo e
    combinar de nao mexer nele nunca mais.
    """
    return _rich(valor)


@dataclass(frozen=True)
class ImagemEmLinha:
    """Uma imagem a embutir no documento, declarada sem depender do docxtpl.

    O `InlineImage` do docxtpl exige o objeto `DocxTemplate`, que so existe
    dentro de `renderizar_modelo`. Se quem monta o contexto precisasse dele, a
    montagem deixaria de ser uma funcao pura sobre dados congelados — e a
    rubrica do instrutor, que e o unico caso hoje (fatia 4 de Certificados),
    passa exatamente por ali. Aqui o contexto declara *qual arquivo*, e a
    conversao acontece no unico lugar que abre o modelo.

    Arquivo ausente vira string vazia, e nao erro: rubrica que sumiu do disco
    nao pode impedir a segunda via de um certificado que ja circulou.
    """

    caminho: Path
    largura_mm: float | None = None


def _resolver_imagens(tpl: DocxTemplate, contexto: dict) -> dict:
    resolvido = {}
    for chave, valor in contexto.items():
        if isinstance(valor, ImagemEmLinha):
            if not valor.caminho.is_file():
                resolvido[chave] = ""
                continue
            largura = Mm(valor.largura_mm) if valor.largura_mm else None
            resolvido[chave] = InlineImage(tpl, str(valor.caminho), width=largura)
        else:
            resolvido[chave] = valor
    return resolvido


def primeira_maiuscula(texto: str | None) -> str:
    """Equivalente ao switch \\* FirstCap da mala direta."""
    if not texto:
        return ""
    return texto[0].upper() + texto[1:]


# ---------------------------------------------------------------------
# Contexto
# ---------------------------------------------------------------------
@dataclass
class ContextoParecer:
    """Tudo o que o modelo precisa. Montado a partir do banco ou dos testes."""

    numero_parecer: int
    ano: int
    data_emissao: date
    cidade: str
    sigla_unidade_emissora: str
    nome_extenso_emissor: str
    endereco_emissor: str
    telefone_emissor: str
    laudo_de: str
    unidade: str
    postos: list[str]
    uorg_bruto: str
    tipo_laudo: str
    numero_processo_sei: str
    nome_servidor: str
    matricula: str
    cargo: str
    funcao: str
    laudo_siape: str
    agentes_nocivos: list[str]
    tipo_risco: str
    percentual_aplicavel: str
    portaria_localizacao: str
    fundamentacao_legal: str
    alteracao: str | None
    recomendacao_tecnica: str
    reavaliacao: str | None
    pro_reitor: str
    pro_reitor_cargo: str
    tratamento_destinatario: str
    assinante_nome: str
    assinante_matricula: str
    assinante_titulo: str
    data_extenso_capitalizado: bool = False
    avisos: list[str] = field(default_factory=list)

    # ---- congelamento (RN-15) ---------------------------------------
    CAMPOS_CONGELADOS = (
        "numero_parecer", "ano", "data_emissao", "cidade", "sigla_unidade_emissora",
        "nome_extenso_emissor", "endereco_emissor", "telefone_emissor", "laudo_de",
        "unidade", "postos", "uorg_bruto", "tipo_laudo", "numero_processo_sei",
        "nome_servidor", "matricula", "cargo", "funcao", "laudo_siape",
        "agentes_nocivos", "tipo_risco", "percentual_aplicavel",
        "portaria_localizacao", "fundamentacao_legal", "alteracao",
        "recomendacao_tecnica", "reavaliacao", "pro_reitor", "pro_reitor_cargo",
        "tratamento_destinatario", "assinante_nome", "assinante_matricula",
        "assinante_titulo", "data_extenso_capitalizado",
    )

    def congelar(self) -> dict:
        """Tudo o que o documento renderiza, pronto para guardar no parecer."""
        dados = {campo: getattr(self, campo) for campo in self.CAMPOS_CONGELADOS}
        dados["data_emissao"] = self.data_emissao.isoformat()
        return dados

    @classmethod
    def descongelar(cls, dados: dict) -> "ContextoParecer":
        from datetime import date as _date

        valores = dict(dados)
        valores["data_emissao"] = _date.fromisoformat(valores["data_emissao"])
        return cls(**{k: valores[k] for k in cls.CAMPOS_CONGELADOS})

    def obrigatorios_faltantes(self) -> list[str]:
        exigidos = {
            "nome_servidor": "nome do servidor",
            "matricula": "matricula SIAPE",
            "unidade": "unidade / UORG",
            "numero_processo_sei": "numero do processo (NUP)",
            "laudo_siape": "laudo tecnico",
            "percentual_aplicavel": "percentual aplicavel",
            "portaria_localizacao": "portaria de localizacao",
            "recomendacao_tecnica": "recomendacao",
            "assinante_nome": "signatario",
            "pro_reitor": "autoridade destinataria",
            "tipo_risco": "tipo de risco",
            "fundamentacao_legal": "fundamentacao legal",
        }
        faltantes = [rotulo for campo, rotulo in exigidos.items() if not getattr(self, campo)]
        if not self.agentes_nocivos:
            faltantes.append("agente nocivo")
        if not self.postos:
            faltantes.append("posto de trabalho")
        return faltantes

    def como_dicionario(self) -> dict:
        data_txt = (
            datas_br.por_extenso_capitalizado(self.data_emissao)
            if self.data_extenso_capitalizado
            else datas_br.por_extenso(self.data_emissao)
        )
        return {
            "numero_parecer": str(self.numero_parecer),
            "ano": str(self.ano),
            "data_extenso": data_txt,
            "cidade": self.cidade,
            "sigla_unidade_emissora": self.sigla_unidade_emissora,
            "nome_extenso_emissor": self.nome_extenso_emissor,
            "endereco_emissor": self.endereco_emissor,
            "telefone_emissor": self.telefone_emissor,
            "laudo_de": self.laudo_de,
            "unidade": primeira_maiuscula(self.unidade),
            "posto_trabalho": _rich("\n".join(self.postos)),
            "uorg_formatado": textos.uorg_formatado(self.uorg_bruto),
            "tipo_laudo": self.tipo_laudo,
            "numero_processo_sei": self.numero_processo_sei,
            "nome_servidor": self.nome_servidor,
            "matricula": self.matricula,
            "cargo": self.cargo,
            "funcao": self.funcao or "",
            "laudo_siape": self.laudo_siape,
            "agente_nocivo": _rich("\n".join(self.agentes_nocivos)),
            "tipo_risco": self.tipo_risco,
            "percentual_aplicavel": self.percentual_aplicavel,
            "portaria_localizacao": self.portaria_localizacao,
            "fundamentacao_legal": _rich(self.fundamentacao_legal),
            "alteracao": _rich(self.alteracao),
            "recomendacao_tecnica": _rich(self.recomendacao_tecnica),
            "reavaliacao": _rich(self.reavaliacao),
            "pro_reitor": self.pro_reitor,
            "pro_reitor_cargo": self.pro_reitor_cargo,
            "tratamento_destinatario": self.tratamento_destinatario,
            "assinante_nome": self.assinante_nome,
            "assinante_matricula": self.assinante_matricula,
            "assinante_titulo": self.assinante_titulo,
            # artefato de mala direta que sobrou no rodape: sempre vazio
            "data_solicitacao_sest_rodape": "",
        }


# ---------------------------------------------------------------------
# Render
# ---------------------------------------------------------------------
def renderizar_modelo(
    modelo: str | None,
    contexto: dict,
    destino: Path,
    *,
    dica: str = "",
    pasta: Path | None = None,
) -> dict:
    """Rende um .docx de `app/templates/` com docxtpl e devolve as impressoes digitais.

    E o unico lugar do sistema que abre um modelo e grava um documento. A lista
    de presenca (fatia 3 de Certificados) e o certificado (fatia 4) passam por
    aqui em vez de trazer cada um o seu render: dois mecanismos de documento
    seriam dois lugares para o hash de integridade divergir.

    `pasta` existe porque o modelo do certificado mora em
    `app/templates/certificados/`, e nao ao lado do modelo do parecer: o setor
    troca o layout do certificado com frequencia e a subpasta e o que impede a
    tela de modelos de enxergar (e oferecer) o `.docx` do parecer. So o NOME do
    arquivo e usado — `Path(...).name` na origem —, entao a pasta nunca vem de
    fora.
    """
    caminho = (pasta / Path(modelo).name) if pasta is not None else caminho_modelo(modelo)
    if not caminho.exists():
        raise FileNotFoundError(
            f"modelo {caminho.name} nao encontrado em {caminho.parent}."
            + (f" {dica}" if dica else "")
        )

    tpl = DocxTemplate(str(caminho))
    tpl.render(_resolver_imagens(tpl, contexto))
    destino.parent.mkdir(parents=True, exist_ok=True)
    tpl.save(str(destino))

    return {
        "arquivo": destino,
        "modelo_arquivo": caminho.name,
        "modelo_sha256": sha256_arquivo(caminho),
        "hash_conteudo": sha256_texto(extrair_texto(destino)),
    }


def renderizar(
    contexto: ContextoParecer,
    destino: Path,
    modelo: str | None = None,
    validar: bool = True,
) -> dict:
    if validar:
        faltantes = contexto.obrigatorios_faltantes()
        if faltantes:
            raise DadosIncompletos(faltantes)

    return renderizar_modelo(
        modelo,
        contexto.como_dicionario(),
        destino,
        dica="Rode ferramentas/converter_modelo_maladireta.py.",
    )


# ---------------------------------------------------------------------
# Extracao de texto (CA-01)
# ---------------------------------------------------------------------
def _texto_de_celula(celula) -> str:
    return "\n".join(p.text for p in celula.paragraphs)


def _texto_de_container(container) -> list[str]:
    from docx.oxml.ns import qn
    from docx.table import Table
    from docx.text.paragraph import Paragraph

    elemento = getattr(container, "element", None)
    if elemento is None:
        elemento = getattr(container, "_element", container)
    elemento = getattr(elemento, "body", elemento)
    linhas: list[str] = []
    for filho in elemento.iterchildren():
        if filho.tag == qn("w:p"):
            linhas.append(Paragraph(filho, container).text)
        elif filho.tag == qn("w:tbl"):
            tabela = Table(filho, container)
            for linha in tabela.rows:
                vistas: set[int] = set()
                for celula in linha.cells:
                    if id(celula._tc) in vistas:
                        continue
                    vistas.add(id(celula._tc))
                    linhas.append(_texto_de_celula(celula))
    return linhas


def extrair_texto(caminho: Path) -> str:
    """Corpo (paragrafos + celulas) + header + footer, nesta ordem."""
    doc = Document(str(caminho))
    linhas = _texto_de_container(doc)
    for secao in doc.sections:
        for parte in (secao.header, secao.footer):
            linhas.extend(_texto_de_container(parte))
    bruto = "\n".join(linhas)
    return unicodedata.normalize("NFC", bruto)


def texto_normalizado(caminho: Path) -> str:
    return textos.normalizar_para_ouro(extrair_texto(caminho))


# ---------------------------------------------------------------------
# Nomes de arquivo
# ---------------------------------------------------------------------
def caminho_saida(
    numero: int, ano: int, sigla: str, servidor: str | None, extensao: str = "docx"
) -> Path:
    cfg = obter_config()
    pasta = cfg.caminho(cfg.dir_documentos) / str(ano)
    nome = textos.nome_arquivo_parecer(numero, ano, sigla, servidor)
    return pasta / f"{nome}.{extensao}"

File: nuvem/src/servicos/documento.ts
/**
 * Geração do parecer técnico (e dos outros documentos) em .docx a partir do
 * modelo. Porte de `app/servicos/documento.py`.
 *
 * O parecer é um .docx de layout fixo com tabelas de células mescladas,
 * cabeçalho MEC/UFVJM com brasão e rodapé de duas colunas. Só o conteúdo
 * varia — o cabeçalho NUNCA é recriado por código.
 *
 * **O motor mudou, o modelo não.** O Python usava o docxtpl (Jinja2 dentro do
 * Word); aqui é o docxtemplater. Os modelos de `app/templates/*.docx` são
 * convertidos UMA vez por `ferramentas/converter-modelos-docx.ts` para
 * `modelos-docx/` (mesmo nome de arquivo), e o `manifesto.json` de lá guarda o
 * SHA-256 do modelo de ORIGEM — é ele que vai para `modelo_sha256`, para que o
 * parecer emitido na nuvem aponte a mesma versão de modelo que o Python
 * apontava. O que o docxtpl fazia depois do Jinja (quebra de linha, parágrafo
 * do RichText, imagem em linha) está em `docx_baixo_nivel.ts`.
 *
 * **Sem disco.** `renderizar_modelo` devolve os bytes; quando recebe
 * `destino`, grava no armazenamento (Supabase Storage / pasta local) com essa
 * chave — o `destino: Path` do Python.
 */
import { createHash } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import Docxtemplater from "docxtemplater";
import PizZip from "pizzip";
import * as datas_br from "./datas_br.js";
import * as textos from "./textos.js";
import { obterArmazenamento } from "./armazenamento.js";
import {
  MARCA_IMAGEM_FIM,
  MARCA_IMAGEM_INICIO,
  MARCA_PARAGRAFO,
  embutir_imagens,
  extrair_texto_docx,
  partes_renderizaveis,
  preparar_imagem,
  resolver_listagem,
  type ImagemPronta,
} from "./docx_baixo_nivel.js";

export const MODELO_V1 = "modelo_parecer_v1.docx";
export const MODELO_V2 = "modelo_parecer_v2.docx";

export const CHAVES_RICHTEXT = [
  "posto_trabalho",
  "agente_nocivo",
  "fundamentacao_legal",
  "alteracao",
  "recomendacao_tecnica",
  "reavaliacao",
] as const;

/** RN-04 — a emissão lista o que falta, nunca falha em silêncio. */
export class DadosIncompletos extends Error {
  constructor(public faltantes: string[]) {
    super("Faltam dados obrigatorios: " + faltantes.join("; "));
  }
}

export class ExposicoesDivergentes extends Error {}

/** O `FileNotFoundError` do modelo ausente. */
export class ModeloNaoEncontrado extends Error {}

export function sha256_arquivo(conteudo: Uint8Array): string {
  return createHash("sha256").update(conteudo).digest("hex");
}

export function sha256_texto(texto: string): string {
  return createHash("sha256").update(texto, "utf8").digest("hex");
}

// ---------------------------------------------------------------------
// Onde moram os modelos convertidos
// ---------------------------------------------------------------------
/**
 * `modelos-docx/` — local em desenvolvimento, e junto do bundle na função do
 * Netlify (tem de estar em `included_files`, como `templates/`).
 */
export function pasta_modelos(): string {
  const candidatos = [
    process.env.CSSO_MODELOS_DOCX,
    path.resolve(process.cwd(), "modelos-docx"),
    path.resolve(process.cwd(), "nuvem/modelos-docx"),
    path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../../modelos-docx"),
    path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../modelos-docx"),
  ].filter(Boolean) as string[];
  for (const c of candidatos) if (existsSync(path.join(c, MODELO_V1))) return c;
  return candidatos[0] ?? path.resolve(process.cwd(), "modelos-docx");
}

interface Manifesto {
  [arquivo: string]: { sha256_origem: string; sha256_convertido: string; marcadores: string[] };
}
let _manifesto: Manifesto | null = null;
function manifesto(): Manifesto {
  if (_manifesto) return _manifesto;
  const arquivo = path.join(pasta_modelos(), "manifesto.json");
  _manifesto = existsSync(arquivo) ? (JSON.parse(readFileSync(arquivo, "utf8")) as Manifesto) : {};
  return _manifesto;
}

/** O nome relativo do modelo (`pasta` é subpasta de `modelos-docx/`, ex. `certificados`). */
export function caminho_modelo(nome: string | null | undefined, pasta?: string | null): string {
  // só o NOME do arquivo é usado — `Path(...).name` na origem —, então a pasta
  // nunca vem de fora
  const arquivo = path.basename(nome || MODELO_V1);
  return pasta ? `${path.basename(pasta)}/${arquivo}` : arquivo;
}

// ---------------------------------------------------------------------
// RichText e imagem
// ---------------------------------------------------------------------
/**
 * O `RichText` do docxtpl, reduzido ao que o sistema usa: linhas separadas
 * por `\a` (parágrafo novo, com o mesmo pPr) dentro de uma célula.
 */
export class RichText {
  readonly partes: string[] = [];
  add(texto: string): this {
    this.partes.push(texto);
    return this;
  }
  /** O valor que atravessa o docxtemplater (o `\a` vira marcador). */
  paraModelo(): string {
    return this.partes.join("").replaceAll("\x07", MARCA_PARAGRAFO);
  }
  toString(): string {
    return this.partes.join("");
  }
}

/** Quebras de linha em célula: RichText com `\a` e `{{r chave }}`. */
function rich(valor: string | null | undefined): RichText | null {
  if (valor === null || valor === undefined) return null;
  const texto = String(valor).replaceAll("\r\n", "\n").replace(/\s+$/u, "");
  if (texto === "") return null;
  const rt = new RichText();
  texto.split("\n").forEach((linha, i) => {
    if (i) rt.add("\x07");
    rt.add(linha);
  });
  return rt;
}

/**
 * `rich` para os outros documentos do sistema (lista de presença,
 * comprovante de EPI, certificado).
 */
export function rich_multilinha(valor: string | null | undefined): RichText | null {
  return rich(valor);
}

/**
 * Uma imagem a embutir no documento, declarada sem depender do motor. No
 * Python era `caminho: Path`; aqui é a chave no armazenamento (ou os bytes).
 * Arquivo ausente vira string vazia, e não erro: rubrica que sumiu não pode
 * impedir a segunda via de um certificado que já circulou.
 */
export class ImagemEmLinha {
  readonly chave: string | null;
  readonly bytes: Uint8Array | null;
  readonly largura_mm: number | null;
  constructor(d: { chave?: string | null; bytes?: Uint8Array | null; largura_mm?: number | null }) {
    this.chave = d.chave ?? null;
    this.bytes = d.bytes ?? null;
    this.largura_mm = d.largura_mm ?? null;
    Object.freeze(this);
  }
  /** O `caminho.is_file()` do Python. */
  async existe(): Promise<boolean> {
    if (this.bytes) return true;
    if (!this.chave) return false;
    return obterArmazenamento().existe(this.chave);
  }
}

/**
 * O contexto como o docxtemplater o lê: RichText vira texto com marcador,
 * ImagemEmLinha vira marcador de imagem (ou "" quando o arquivo sumiu),
 * `null` vira "" (ver DESVIOS: o Jinja imprimiria "None").
 */
async function resolver_contexto(contexto: Record<string, unknown>, imagens: ImagemPronta[]): Promise<Record<string, unknown>> {
  const resolver = async (valor: unknown): Promise<unknown> => {
    if (valor instanceof RichText) return valor.paraModelo();
    if (valor instanceof ImagemEmLinha) {
      let bytes = valor.bytes;
      if (!bytes && valor.chave) {
        if (!(await obterArmazenamento().existe(valor.chave))) return "";
        bytes = await obterArmazenamento().ler(valor.chave);
      }
      if (!bytes) return "";
      imagens.push(preparar_imagem(bytes, valor.largura_mm));
      return `${MARCA_IMAGEM_INICIO}${imagens.length - 1}${MARCA_IMAGEM_FIM}`;
    }
    if (Array.isArray(valor)) return Promise.all(valor.map(resolver));
    if (valor && typeof valor === "object" && !(valor instanceof Date)) {
      const saida: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(valor)) saida[k] = await resolver(v);
      return saida;
    }
    if (valor === null || valor === undefined) return "";
    return valor;
  };
  return (await resolver(contexto)) as Record<string, unknown>;
}

export function primeira_maiuscula(texto: string | null | undefined): string {
  if (!texto) return "";
  const [p, ...resto] = [...texto];
  return p!.toUpperCase() + resto.join("");
}

// ---------------------------------------------------------------------
// Contexto
// ---------------------------------------------------------------------
export interface DadosContextoParecer {
  numero_parecer: number;
  ano: number;
  /** 'AAAA-MM-DD' */
  data_emissao: string;
  cidade: string;
  sigla_unidade_emissora: string;
  nome_extenso_emissor: string;
  endereco_emissor: string;
  telefone_emissor: string;
  laudo_de: string;
  unidade: string;
  postos: string[];
  uorg_bruto: string;
  tipo_laudo: string;
  numero_processo_sei: string;
  nome_servidor: string;
  matricula: string;
  cargo: string;
  funcao: string;
  laudo_siape: string;
  agentes_nocivos: string[];
  tipo_risco: string;
  percentual_aplicavel: string;
  portaria_localizacao: string;
  fundamentacao_legal: string;
  alteracao: string | null;
  recomendacao_tecnica: string;
  reavaliacao: string | null;
  pro_reitor: string;
  pro_reitor_cargo: string;
  tratamento_destinatario: string;
  assinante_nome: string;
  assinante_matricula: string;
  assinante_titulo: string;
  data_extenso_capitalizado?: boolean;
  avisos?: string[];
}

/** Tudo o que o modelo precisa. Montado a partir do banco ou dos testes. */
export class ContextoParecer implements DadosContextoParecer {
  numero_parecer!: number;
  ano!: number;
  data_emissao!: string;
  cidade!: string;
  sigla_unidade_emissora!: string;
  nome_extenso_emissor!: string;
  endereco_emissor!: string;
  telefone_emissor!: string;
  laudo_de!: string;
  unidade!: string;
  postos!: string[];
  uorg_bruto!: string;
  tipo_laudo!: string;
  numero_processo_sei!: string;
  nome_servidor!: string;
  matricula!: string;
  cargo!: string;
  funcao!: string;
  laudo_siape!: string;
  agentes_nocivos!: string[];
  tipo_risco!: string;
  percentual_aplicavel!: string;
  portaria_localizacao!: string;
  fundamentacao_legal!: string;
  alteracao!: string | null;
  recomendacao_tecnica!: string;
  reavaliacao!: string | null;
  pro_reitor!: string;
  pro_reitor_cargo!: string;
  tratamento_destinatario!: string;
  assinante_nome!: string;
  assinante_matricula!: string;
  assinante_titulo!: string;
  data_extenso_capitalizado = false;
  avisos: string[] = [];

  constructor(d: DadosContextoParecer) {
    Object.assign(this, d);
    this.data_extenso_capitalizado = d.data_extenso_capitalizado ?? false;
    this.avisos = d.avisos ?? [];
  }

  // ---- congelamento (RN-15) ---------------------------------------
  static readonly CAMPOS_CONGELADOS = [
    "numero_parecer", "ano", "data_emissao", "cidade", "sigla_unidade_emissora",
    "nome_extenso_emissor", "endereco_emissor", "telefone_emissor", "laudo_de",
    "unidade", "postos", "uorg_bruto", "tipo_laudo", "numero_processo_sei",
    "nome_servidor", "matricula", "cargo", "funcao", "laudo_siape",
    "agentes_nocivos", "tipo_risco", "percentual_aplicavel",
    "portaria_localizacao", "fundamentacao_legal", "alteracao",
    "recomendacao_tecnica", "reavaliacao", "pro_reitor", "pro_reitor_cargo",
    "tratamento_destinatario", "assinante_nome", "assinante_matricula",
    "assinante_titulo", "data_extenso_capitalizado",
  ] as const;

  /** Tudo o que o documento renderiza, pronto para guardar no parecer. */
  congelar(): Record<string, unknown> {
    const dados: Record<string, unknown> = {};
    for (const campo of ContextoParecer.CAMPOS_CONGELADOS) {
      const v = this[campo];
      dados[campo] = Array.isArray(v) ? [...v] : v;
    }
    // a data já é 'AAAA-MM-DD' (o `isoformat()` do Python)
    dados.data_emissao = datas_br.comoData(this.data_emissao);
    return dados;
  }

  static descongelar(dados: Record<string, unknown>): ContextoParecer {
    const valores: Record<string, unknown> = {};
    for (const k of ContextoParecer.CAMPOS_CONGELADOS) {
      if (!(k in dados)) throw new Error(`campo congelado ausente: ${k}`); // o KeyError do Python
      valores[k] = dados[k];
    }
    valores.data_emissao = datas_br.comoData(valores.data_emissao);
    return new ContextoParecer(valores as unknown as DadosContextoParecer);
  }

  obrigatorios_faltantes(): string[] {
    const exigidos: [keyof DadosContextoParecer, string][] = [
      ["nome_servidor", "nome do servidor"],
      ["matricula", "matricula SIAPE"],
      ["unidade", "unidade / UORG"],
      ["numero_processo_sei", "numero do processo (NUP)"],
      ["laudo_siape", "laudo tecnico"],
      ["percentual_aplicavel", "percentual aplicavel"],
      ["portaria_localizacao", "portaria de localizacao"],
      ["recomendacao_tecnica", "recomendacao"],
      ["assinante_nome", "signatario"],
      ["pro_reitor", "autoridade destinataria"],
      ["tipo_risco", "tipo de risco"],
      ["fundamentacao_legal", "fundamentacao legal"],
    ];
    const faltantes = exigidos.filter(([campo]) => !this[campo]).map(([, rotulo]) => rotulo);
    if (!this.agentes_nocivos?.length) faltantes.push("agente nocivo");
    if (!this.postos?.length) faltantes.push("posto de trabalho");
    return faltantes;
  }

  como_dicionario(): Record<string, unknown> {
    const data_txt = this.data_extenso_capitalizado
      ? datas_br.por_extenso_capitalizado(this.data_emissao)
      : datas_br.por_extenso(this.data_emissao);
    return {
      numero_parecer: String(this.numero_parecer),
      ano: String(this.ano),
      data_extenso: data_txt,
      cidade: this.cidade,
      sigla_unidade_emissora: this.sigla_unidade_emissora,
      nome_extenso_emissor: this.nome_extenso_emissor,
      endereco_emissor: this.endereco_emissor,
      telefone_emissor: this.telefone_emissor,
      laudo_de: this.laudo_de,
      unidade: primeira_maiuscula(this.unidade),
      posto_trabalho: rich(this.postos.join("\n")),
      uorg_formatado: textos.uorg_formatado(this.uorg_bruto),
      tipo_laudo: this.tipo_laudo,
      numero_processo_sei: this.numero_processo_sei,
      nome_servidor: this.nome_servidor,
      matricula: this.matricula,
      cargo: this.cargo,
      funcao: this.funcao || "",
      laudo_siape: this.laudo_siape,
      agente_nocivo: rich(this.agentes_nocivos.join("\n")),
      tipo_risco: this.tipo_risco,
      percentual_aplicavel: this.percentual_aplicavel,
      portaria_localizacao: this.portaria_localizacao,
      fundamentacao_legal: rich(this.fundamentacao_legal),
      alteracao: rich(this.alteracao),
      recomendacao_tecnica: rich(this.recomendacao_tecnica),
      reavaliacao: rich(this.reavaliacao),
      pro_reitor: this.pro_reitor,
      pro_reitor_cargo: this.pro_reitor_cargo,
      tratamento_destinatario: this.tratamento_destinatario,
      assinante_nome: this.assinante_nome,
      assinante_matricula: this.assinante_matricula,
      assinante_titulo: this.assinante_titulo,
      // artefato de mala direta que sobrou no rodapé: sempre vazio
      data_solicitacao_sest_rodape: "",
    };
  }
}

// ---------------------------------------------------------------------
// Render
// ---------------------------------------------------------------------
export interface DocumentoRenderizado {
  /** o .docx pronto */
  bytes: Uint8Array;
  /** a chave no armazenamento, quando `destino` foi pedido (o `destino` do Python) */
  arquivo: string | null;
  modelo_arquivo: string;
  modelo_sha256: string;
  hash_conteudo: string;
}

/** Os bytes do modelo convertido e o SHA-256 que o identifica. */
export function carregar_modelo(
  modelo: string | null | undefined,
  opcoes: { pasta?: string | null; dica?: string } = {},
): { bytes: Uint8Array; arquivo: string; sha256: string } {
  const relativo = caminho_modelo(modelo, opcoes.pasta);
  const absoluto = path.join(pasta_modelos(), relativo);
  if (!existsSync(absoluto)) {
    throw new ModeloNaoEncontrado(
      `modelo ${path.basename(relativo)} nao encontrado em ${path.dirname(absoluto)}.` +
        (opcoes.dica ? ` ${opcoes.dica}` : ""),
    );
  }
  const bytes = new Uint8Array(readFileSync(absoluto));
  // o SHA-256 do modelo de ORIGEM (docxtpl), o mesmo que o Python registrava
  const sha256 = manifesto()[relativo]?.sha256_origem ?? sha256_arquivo(bytes);
  return { bytes, arquivo: path.basename(relativo), sha256 };
}

/**
 * Rende um .docx de `modelos-docx/` e devolve as impressões digitais.
 *
 * É o único lugar do sistema que abre um modelo e gera um documento: a lista
 * de presença e o certificado passam por aqui — dois mecanismos de documento
 * seriam dois lugares para o hash de integridade divergir.
 */
export async function renderizar_modelo(
  modelo: string | null | undefined,
  contexto: Record<string, unknown>,
  opcoes: { destino?: string | null; dica?: string; pasta?: string | null } = {},
): Promise<DocumentoRenderizado> {
  const origem = carregar_modelo(modelo, opcoes);
  const bytes = await renderizar_bytes(origem.bytes, contexto);
  if (opcoes.destino) {
    await obterArmazenamento().gravar(
      opcoes.destino,
      bytes,
      "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    );
  }
  return {
    bytes,
    arquivo: opcoes.destino ?? null,
    modelo_arquivo: origem.arquivo,
    modelo_sha256: origem.sha256,
    hash_conteudo: sha256_texto(extrair_texto(bytes)),
  };
}

/** O render em si, sobre os bytes de um modelo JÁ convertido. */
export async function renderizar_bytes(modelo: Uint8Array, contexto: Record<string, unknown>): Promise<Uint8Array> {
  const imagens: ImagemPronta[] = [];
  const dados = await resolver_contexto(contexto, imagens);
  const zip = new PizZip(modelo);
  const doc = new Docxtemplater(zip, {
    paragraphLoop: true,
    // variável ausente sai vazia, como o `Undefined` do Jinja
    nullGetter: () => "",
  });
  doc.render(dados);
  const saida = doc.getZip() as PizZip;
  for (const parte of partes_renderizaveis(saida)) {
    const xml = saida.file(parte)!.asText();
    const resolvido = resolver_listagem(xml);
    if (resolvido !== xml) saida.file(parte, resolvido);
    embutir_imagens(saida, parte, imagens);
  }
  return saida.generate({ type: "uint8array", compression: "DEFLATE" });
}

export async function renderizar(
  contexto: ContextoParecer,
  destino: string | null = null,
  modelo: string | null = null,
  validar = true,
): Promise<DocumentoRenderizado> {
  if (validar) {
    const faltantes = contexto.obrigatorios_faltantes();
    if (faltantes.length) throw new DadosIncompletos(faltantes);
  }
  return renderizar_modelo(modelo, contexto.como_dicionario(), {
    destino,
    dica: "Rode ferramentas/converter-modelos-docx.ts.",
  });
}

// ---------------------------------------------------------------------
// Extração de texto (CA-01)
// ---------------------------------------------------------------------
/** Corpo (parágrafos + células) + header + footer, nesta ordem. */
export function extrair_texto(docx: Uint8Array): string {
  return extrair_texto_docx(docx);
}

export function texto_normalizado(docx: Uint8Array): string {
  return textos.normalizar_para_ouro(extrair_texto(docx));
}

// ---------------------------------------------------------------------
// Nomes de arquivo
// ---------------------------------------------------------------------
/** A chave do documento no armazenamento: `documentos/<ano>/<nome>.<ext>`. */
export function caminho_saida(
  numero: number,
  ano: number,
  sigla: string,
  servidor: string | null | undefined,
  extensao = "docx",
): string {
  return `documentos/${ano}/${textos.nome_arquivo_parecer(numero, ano, sigla, servidor)}.${extensao}`;
}
